### PR TITLE
fix(cayenne): keep a file's statistics the same whichever source serves them (refs #13829)

### PR DESCRIPTION
**Pull request by claudespice**
## Summary

A Cayenne file scan has two sources for one file's statistics, and they did not agree.
`CayenneTableProvider::collect_scan_file_statistics` serves either the Vortex footer
(`VortexFormat::infer_stats`) or the `cayenne_snapshot_file_statistics` blob that the footer
path itself writes on its way out. The footer reports each column's
`UncompressedSizeInBytes` as `ColumnStatistics::byte_size` and their sum as
`total_byte_size`; the blob carried neither, because `column_stats_to_stats_set` persisted
only min/max/sum/null-count and `stats_set_to_column_stats` hard-coded
`byte_size: Precision::Absent` beside a hard-coded `total_byte_size: Precision::Absent`.

`JoinSelection` compares `total_byte_size` before anything else, so the same file answered a
size or answered nothing depending on which source happened to serve it — which flips with a
restart (after one, every file takes the blob path), with promotion (cold files are *only*
ever served from the blob), and between two scans of one plan that race to write it. Results
were never wrong; plan stability was. #13829 attributes the recurring Cayenne
explain-snapshot churn (#13774 and #13808 re-baselining the same joins in opposite
directions) to this — that attribution is the issue's, carried here rather than re-measured.

## Changes

- **`crates/cayenne/src/stats.rs`** — persist the per-column byte size into the blob as
  `Stat::UncompressedSizeInBytes`, read it back into `ColumnStatistics::byte_size`, and derive
  `total_byte_size` by summing the per-column sizes the way `infer_stats` does. `Precision::add`
  absorbs `Absent` exactly as Vortex's `Precision::zip` does in `infer_stats`, so one column
  without a size makes the total `Absent` rather than a partial sum that would report a file as
  a fraction of its real size.
- **`crates/cayenne/src/provider/table.rs`** — a blob written before the sizes were persisted
  has no total, so it is treated as stale and re-inferred from the footer, which rewrites it
  with the size in it. Without this the fix would only reach files written after it.
- **`docs/cayenne/cayenne.md`** — the per-file statistics paragraph said the blob carries
  "min/max/null/row-count", and the restore paragraph said raw Vortex data is never re-read.
  Both are now true of the code again.

The converter is shared with the table-level accumulator, which is why the change is written
to be inert there: `ColumnStatsAccumulator::compute_column_stats` (`provider/column_stats.rs`) sets
`byte_size: Absent`
unconditionally, so no size is ever offered on that path, nothing new is persisted, and the
Vortex `merge_unordered` that combines those blobs across writes sees exactly what it saw
before.

## Test plan
- [x] Added regression test for a file's statistics differing between the footer source and the persisted-blob source across a restart
- [x] Added edge case tests for a legacy blob carrying no sizes, and for a partial blob where only some columns carry one
- [x] Added an end-to-end migration test that seeds a pre-change row, reopens the table, and asserts the footer size is served **and** the row rewritten (Copilot review)
- [x] `make lint` passes (workspace-wide `cargo fmt --all -- --check` + full clippy — not a per-crate scoped invocation): both clippy passes completed with **0 errors** in the local `make signoff` run before it died on disk, and again in remote run `33942994543` before its wall
- [x] `make test` passes — via the green remote sign-off below (`make nextest verify-cli`). `cayenne` in full is also green locally: **1663 passed / 0 failed across 73 binaries**
- [x] `make signoff` run after the final push (`Attestation` check is green) — run `33984719072`, `lint+test passed in 9351s`
- [ ] <no query-plan snapshot tests apply>

### Sign-off status — green on the seventh attempt, after six infrastructure failures

| attempt | where | outcome |
|---|---|---|
| local | this workstation | `No space left on device` compiling `data_components` test targets — **after** both clippy passes returned 0 errors and the targeted pre-check ran 1211/1211. The guard correctly posted no status. |
| `33940183941` | `github-runner-03`, 3431 s | ``could not execute process `sccache …` (never executed)`` → `No such file or directory`. The compiler never ran; failing units were third-party (`asn1-rs-derive`, `turso_core`). |
| `33942994543` | 21203 s | `The action 'Sign off' has timed out after 353 minutes` — the step wall. 0 errors in 5437 log lines. |
| `33958948484` | — | `cancelled`, correctly superseded by the review-fix push. |
| `33962205883` | 21221 s | The wall again — but this one **also** caught a real flake of my own: `TRY 1 FAIL` / `TRY 2 PASS` on both backends at the `!files.is_empty()` premise. Fixed by polling the condition with a bound (`8cfbb8e`), which is the same defect filed as [#13906](https://github.com/spiceai/spiceai/issues/13906) against a neighbouring suite. |
| `33984719072` | 9351 s | **success** — `lint+test passed`. |

## Reproduction

`crates/cayenne/tests/file_scan_statistics_source_test.rs` lands rows in a durable Vortex file
and scans (footer path), then reopens the table over the same metastore — the restart case —
and scans again (blob path), comparing the two.

On `trunk`, the same file reports two different sizes:

```
$ cargo test --package cayenne --test file_scan_statistics_source_test
test file_scan_byte_size_statistics_do_not_depend_on_their_source_sqlite ... FAILED

assertion `left == right` failed: the same file must report the same total byte size whichever source served it
  left: Absent
 right: Exact(16274)
```

On this branch, same command:

```
test file_scan_byte_size_statistics_do_not_depend_on_their_source_sqlite ... ok
test result: ok. 1 passed; 0 failed
```

The test asserts the footer path reported a size at all before comparing, so a setup that
stopped producing a file scan would fail loudly rather than pass by agreeing on `Absent`.

**Neuter-verified, both halves independently.** Reverting only the restore side
(`byte_size` back to `Absent`) and, separately, only the persist side, each fails
`per_column_byte_size_survives_the_persisted_blob` with `left: Absent, right: Exact(32)`,
while the four pre-existing `stats::` tests stay green — so neither half is redundant and
neither is what makes the pre-existing tests pass.

## Known residue

- **The single-flight collection the issue also proposes is not here.** With both sources
  producing the same value a concurrent race no longer changes the answer — only how many
  footer reads happen — but that is reasoning, not a measurement: no concurrency experiment
  was run. #13829 stays open owning that decision, which is why this is `refs` and not `fixes`.
- **A cold-tier file whose blob predates this change keeps reporting no size.** The staleness
  re-infer is on the warm path, which has a footer to fall back to; the cold listing path
  (`table.rs`'s cold branch, `cold_partition.rs`) reads the blob only. Cold blobs written after
  this change carry sizes, so this drains on the next promotion rather than persisting.
- **A column in the table schema but absent from an older file** is skipped by `infer_stats`'s
  sum but is indistinguishable in the blob from a column whose size was not recorded, so the
  blob yields `Absent` where the footer yields a total. The staleness guard means such a blob
  is never served on the warm path — the footer answer is — so the statistic a scan sees still
  agrees; the cost is one footer read per such file per process, held afterwards by
  `scan_file_statistics`.

## Performance impact

One extra `u64` stat per column in each per-file blob, and a fold over the columns on restore.
The staleness re-infer costs one footer read per pre-existing file, once per process, after
which the rewritten blob serves. Against that, files that previously reported no size now
report one, so `JoinSelection` picks build sides from real numbers rather than from whichever
source answered.

## Review gate

- Adversarial review: **skipped** — `codex` returned "Your workspace is out of credits" and produced no review (attempted on the working tree and again on the committed head); `grok` is not installed on this host, so `engine=none`. The local receipt records `engine=codex exit=0`, which is the companion scoring a failed turn as a clean process exit — no review was produced, and this line says so rather than letting that receipt read as a pass.
- `/security-review`: no findings — the diff moves only internally-produced statistics, adds no parser or new deserialization surface, and its one new log line emits a byte count. The `Exact`-substitution hazard it surfaced is a correctness concern, handled by degrading a partial sum to `Absent` and covered by `a_missing_column_byte_size_leaves_the_total_absent`.
- `/simplify`: applied — `Precision::add` is absorbing on `Absent`, so a `try_fold` with an `Err` sentinel collapsed to a plain `fold`; light checks re-run green.
- Copilot: **3 findings, all valid, all fixed** in `85869d4` — the migration branch had no end-to-end coverage (two of the three), and the doc claimed a narrower fallback than the code performs. Threads replied to and resolved.

**Substitute for the missing adversarial pass.** Rather than skip the challenge entirely, the
approach's load-bearing assumption was re-derived from source: that summing per-column sizes
reproduces what `infer_stats` computes. Vortex's `Precision::zip`
(`vortex-array/src/expr/stats/precision.rs:144`) returns `Absent` if either side is `Absent`,
exactly as DataFusion's `Precision::add` does, so the accumulators agree — **except** for a
column that `infer_stats` skips because it is not in the file, which `continue`s before the
sum. That case is indistinguishable in the blob from a column whose size was never recorded,
and it is the third entry under *Known residue*; the staleness guard is what keeps it from
becoming a divergence a scan can observe.

**A negative result worth recording.** While building the migration test it looked as though
the persist branch were dead — the seeded row read back unchanged after a scan. Probing the
production path directly disproved it: `per_col=[Exact(4096), Exact(12178)]`,
`roundtrip_total=Some(Exact(16274))`, `upsert_ok`. The test was seeding the wrong key, not the
code failing to write. Two keying details caused it, both now recorded in the test: a per-file
statistics row is keyed by the **object-store location** while the manifest carries the bare
filename, and the scan's `snapshot_id` is the manifest's snapshot-directory id, not the
provider's in-memory `current_snapshot_id()`.

**Unrelated flake found while running the suite**, filed rather than fixed here:
[#13906](https://github.com/spiceai/spiceai/issues/13906) — `small_files_compaction_test`'s
premise assertion is decided by a bounded wait on debounced maintenance, so it fails under
whole-crate load (failed once, then the identical tree and command passed 1663/73, and the test
passes 3/3 in isolation).

Refs #13829

### CODE REVIEW

**Comment by claudespice** [comment]
@copilot review

File: crates/cayenne/src/provider/table.rs [hunk 1]
```diff
@@ -31953,6 +31953,14 @@ impl CayenneTableProvider {
...
                 persisted.num_rows,
             )
+            // A blob written before per-column byte sizes were persisted carries no
+            // total, so serving it would report a different size for this file than
+            // the footer does and leave `JoinSelection` picking a build side by which
+            // source happened to answer. Treat it as stale and re-infer from the
+            // footer below, which rewrites the blob with the size in it. A file whose
+            // footer carries no statistics at all re-infers on first touch in each
+            // process and is then held by `scan_file_statistics`.
+            && statistics.total_byte_size != DFPrecision::Absent
```
[the remark this answers is not in the document]
Comment (claudespice) [reply crates/cayenne/src/provider/table.rs:31963]: Valid, and this was the right thing to insist on — the branch that makes this fix reach *existing* installations was the one piece with no end-to-end coverage.

Added `a_blob_without_byte_sizes_is_re_inferred_from_its_footer` (85869d4): it seeds a `cayenne_snapshot_file_statistics` row shaped like one written before the sizes were persisted, reopens the table over that metastore, and asserts both halves you named — the scan serves the footer's size, and the row is rewritten so the next process does not re-infer again. Neuter-verified: removing the `total_byte_size != Absent` gate fails it with `left: Absent, right: Exact(16274)`.

Writing it took four wrong attempts, and two of them were wrong about *this* file's keying rather than about the fix — worth recording here because the next person will hit them:

* a per-file statistics row is keyed by the **object-store location**; `SnapshotFile::file_path` from the manifest is the **bare filename**, so seeding with the manifest value writes a row the scan never reads;
* the `snapshot_id` the scan passes is the **snapshot-directory id from the manifest**, not the provider's in-memory `current_snapshot_id()` — those are different values.

With both keys wrong the seed and the provider were reading two different rows, which made it look as though the upsert were dead. It is not: probing the production path directly showed `per_col=[Exact(4096), Exact(12178)]`, `roundtrip_total=Some(Exact(16274))`, `upsert_ok`. Reporting that as a negative result rather than "fixing" a defect that was never there.

File: crates/cayenne/tests/file_scan_statistics_source_test.rs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    let reopened = Arc::new(
+        CayenneTableProviderBuilder::new(
+            Arc::clone(&catalog) as Arc<dyn MetadataCatalog>,
+            ctx.runtime_env(),
+        )
+        .open(TABLE)
+        .await?,
+    );
+
+    let (blob_total, blob_columns) = scan_statistics(&reopened, &ctx).await?;
```
[the remark this answers is not in the document]
Comment (claudespice) [reply crates/cayenne/tests/file_scan_statistics_source_test.rs:140]: Same finding as the `table.rs` thread, and correct on both counts — the reopened scan there consumes a blob the *updated* serializer wrote, so it never reaches the legacy path, and the unit case stops at proving the conversion returns `Absent`.

Fixed in 85869d4 by the new `a_blob_without_byte_sizes_is_re_inferred_from_its_footer` in this same file, which installs a blob with no byte sizes, reopens, scans, and asserts both the footer re-inference and the rewritten row. Neuter-verified against the staleness gate (`left: Absent, right: Exact(16274)` with it removed), so it fails if that wiring regresses rather than passing vacuously.

`statistics_to_persisted_blob` is now `pub` so the test can build the seed blob; it is the serializing counterpart of the already-public `deserialize_file_statistics` and `file_statistics_to_df`.

File: docs/cayenne/cayenne.md [hunk 3]
```diff
@@ -756,11 +756,11 @@ There is one place where "inexact, so only costing" was *not* the whole story: t
...
 
-Restore is a metastore read, not a rescan. At open, `load_table_statistics` reads the aggregate blob (`get_table_statistics`) and deserializes it straight into a `Statistics`, and the per-file layer comes back from `cayenne_snapshot_file_statistics`. The raw Vortex data is never re-read to rebuild statistics.
+Restore is a metastore read, not a rescan. At open, `load_table_statistics` reads the aggregate blob (`get_table_statistics`) and deserializes it straight into a `Statistics`, and the per-file layer comes back from `cayenne_snapshot_file_statistics`. The raw Vortex data is never re-read to rebuild statistics — with one bounded exception: a per-file blob written before the byte sizes were persisted carries no total, and is re-inferred from that file's footer once per process rather than served at a size the footer disagrees with.
```
[the remark this answers is not in the document]
Comment (claudespice) [reply docs/cayenne/cayenne.md]: Correct, and the doc contradicted this PR's own *Known residue* section, which already admitted the schema-evolved case — so the narrower wording was wrong twice over.

Reworded in 85869d4 to describe the predicate the code actually evaluates: a warm-path blob whose `total_byte_size` is absent is re-inferred, which covers a pre-change blob, a file whose footer carries no size statistics at all, and a file missing a column the table schema has. It also now states that each re-infers once per process and is then held by the in-memory cache, and that the cold listing path has no footer to fall back to, so it serves such a blob as-is until the next promotion rewrites it.

**Review by lukekim — APPROVED** [review]
